ocaml/idl/datamodel.ml

```diff
@@ -3891,16 +3892,120 @@ let host_crashdump =
-          "restartPV",   "This patch requires PV guests to be restarted once applied.";
-		  "restartHost", "This patch requires the host to be restarted once applied."; 
-          "restartXAPI", "This patch requires XAPI to be restarted once applied.";
+        [ "restartHVM",  "This update requires HVM guests to be restarted once applied.";
+          "restartPV",   "This update requires PV guests to be restarted once applied.";
+		  "restartHost", "This update requires the host to be restarted once applied."; 
+          "restartXAPI", "This update requires XAPI to be restarted once applied.";
         ])
 
+let pool_update_introduce = call
+  ~name:"introduce"
+  ~doc:"Introduce update VDI"
+  ~in_oss_since:None
+  ~in_product_since:rel_ely
+  ~params:[ Ref _vdi, "vdi", "The VDI which contains update" ]
+  ~result:(Ref _pool_update, "the introduced pool update")
+  ~allowed_roles:_R_POOL_OP
+  ()
+
+let pool_update_precheck = call
+  ~name:"precheck"
+  ~doc:"Execute the precheck stage of the selected update on a host"
+  ~in_oss_since:None
+  ~in_product_since:rel_ely
+  ~params:[ Ref _pool_update, "self", "The update whose prechecks will be run"; Ref _host, "host", "The host to run the prechecks on" ]  
+  ~allowed_roles:_R_POOL_OP
+  ~forward_to:(HostExtension "Pool_update.precheck")
+  ()
+
+let pool_update_apply = call
+  ~name:"apply"
+  ~doc:"Apply the selected update to a host and return its output"
+  ~in_oss_since:None
+  ~in_product_since:rel_ely
+  ~params:[ Ref _pool_update, "self", "The update to apply"; Ref _host, "host", "The host to apply the update too" ]
+  ~allowed_roles:_R_POOL_OP
+  ~forward_to:(HostExtension "Pool_update.apply")
+  ()
+
+let pool_update_pool_apply = call
+  ~name:"pool_apply"
+  ~doc:"Apply the selected update to all hosts in the pool and return a map of host_ref -> update output"
+  ~in_oss_since:None
+  ~in_product_since:rel_ely
+  ~params:[ Ref _pool_update, "self", "The update to apply"]
+  ~allowed_roles:_R_POOL_OP
+  ()
+
+let pool_update_clean = call
+  ~name:"clean"
+  ~doc:"Removes the update's files from the server"
+  ~in_oss_since:None
+  ~in_product_since:rel_ely
+  ~params:[ Ref _pool_update, "self", "The update to clean up"; Ref _host, "host", "The host to clean the update too" ] 
+  ~allowed_roles:_R_POOL_OP
+  ()
+
+let pool_update_pool_clean = call
+  ~name:"pool_clean"
+  ~doc:"Removes the update's files from all hosts in the pool, but does not remove the database entries"
+  ~in_oss_since:None
+  ~in_product_since:rel_ely
+  ~params:[ Ref _pool_update, "self", "The update to clean up" ]
+  ~allowed_roles:_R_POOL_OP
+  ()
+
+let pool_update_destroy = call
+  ~name:"destroy"
+  ~doc:"Removes the update's files from all hosts in the pool, and removes the database entries.  Only works on unapplied updates."
+  ~in_oss_since:None
+  ~in_product_since:rel_ely
+  ~params:[ Ref _pool_update, "self", "The update to destroy" ]
+  ~allowed_roles:_R_POOL_OP
+  ()
+
```

Comment:
The design talks about (internal) `attach` and `detach` functions as well. Will you add those in a later patch?